### PR TITLE
KeyAuthAI: Keystroke Dynamics-Based User Authentication

### PR DESCRIPTION
**Pull request by SK8-infi**
<p><strong>Title:</strong><br>
<code inline="">[Code Addition Request]: KeyAuthAI: Keystroke Dynamics-Based User Authentication</code></p>
<hr>
<p><strong>Have you completed your first issue?</strong><br>
☑ I have completed my first issue</p>
<p><strong>Guidelines:</strong><br>
☑ I have read the guidelines</p>
<p><strong>Latest Merged PR Link:</strong><br>
<code inline="">#1567</code></p>
<hr>
<h2><strong>Project Description</strong></h2>
<p>🕵️ <strong>KeyAuthAI: Keystroke Dynamics-Based User Authentication</strong></p>
<h3>📌 <strong>Project Overview</strong></h3>
<p>KeyAuthAI is a <strong>behavioral biometrics authentication system</strong> that identifies users by their <strong>typing rhythm</strong>—including dwell time (key hold duration), flight time (time between key presses), and overall speed patterns.<br>
It creates a <strong>unique profile</strong> for each user and authenticates by comparing real-time typing to stored profiles, using either <strong>supervised</strong> or <strong>unsupervised</strong> machine learning models.</p>
<p>✅ No external datasets required — all data is collected locally during registration.</p>
<hr>
<h3>💡 <strong>Use Cases</strong></h3>
<ul>
<li>
<p>🛡️ <strong>Invisible 2FA</strong> for critical systems</p>
</li>
<li>
<p>🧠 <strong>Behavioral profiling</strong> for fraud detection</p>
</li>
<li>
<p>🖥️ <strong>Continuous authentication</strong> in enterprise environments</p>
</li>
</ul>
<hr>
<h3>📂 <strong>Folder Structure</strong></h3>
<pre><code>keyauthai/
│
├── data/
│   ├── keystroke_logger.py       # Captures timing &amp; key events
│   └── user_data.json            # Stores user profiles &amp; metrics
│
├── features/
│   └── extractor.py              # Calculates dwell, flight, n-gram stats
│
├── model/
│   ├── train_model.py            # Train ML classifier/anomaly detector
│   ├── model_svm.pkl             # Saved model file
│   └── verify_user.py            # Authentication logic
│
├── ui/
│   └── auth_terminal.py          # CLI/GUI for user login
│
├── utils/
│   └── metrics.py                # FAR, FRR, EER, ROC plotting
│
├── tests/
│   └── test_data_collection.py
│
├── requirements.txt
├── README.md
└── LICENSE
</code></pre>
<hr>
<h3>🧠 <strong>What It Measures</strong></h3>

Metric | Description
-- | --
Dwell Time | How long a key is pressed
Flight Time | Time between releasing one key & pressing another
Typing Speed | Characters per second
N-gram Patterns | Frequency of key sequences like "th", "ing"

<hr>
<h3>🔐 <strong>Authentication Flow</strong></h3>
<p><strong>Registration Phase:</strong></p>
<ul>
<li>
<p>User types a fixed passphrase 5–10 times</p>
</li>
<li>
<p>Model learns statistical patterns from the collected data</p>
</li>
</ul>
<p><strong>Verification Phase:</strong></p>
<ul>
<li>
<p>User types the passphrase once</p>
</li>
<li>
<p>Model compares to stored profile &amp; verifies authenticity</p>
</li>
</ul>
<hr>
<h3>🧠 <strong>Model Options</strong></h3>
<ul>
<li>
<p><strong>Supervised:</strong> SVM, KNN, Random Forest</p>
</li>
<li>
<p><strong>Unsupervised:</strong> One-Class SVM, Isolation Forest</p>
</li>
<li>
<p><strong>Advanced:</strong> LSTM for sequential pattern learning</p>
</li>
</ul>
<hr>
<h3>🧪 <strong>Metrics</strong></h3>
<ul>
<li>
<p><strong>FAR:</strong> False Acceptance Rate</p>
</li>
<li>
<p><strong>FRR:</strong> False Rejection Rate</p>
</li>
<li>
<p><strong>EER:</strong> Equal Error Rate (optimization target)</p>
</li>
</ul>
<hr>
<h3>🔧 <strong>Features</strong></h3>
<ul>
<li>
<p>Local biometric authentication</p>
</li>
<li>
<p>Real-time typing pattern collection</p>
</li>
<li>
<p>Lightweight &amp; pluggable into existing login systems</p>
</li>
<li>
<p>No passwords stored — <strong>only typing behavior</strong></p>
</li>
</ul>
<hr>
<h3>📘 <strong>README Highlights</strong></h3>
<p><strong>How it Works:</strong></p>
<ol>
<li>
<p>Capture keystroke timing data while the user types</p>
</li>
<li>
<p>Extract features (dwell, flight, n-grams)</p>
</li>
<li>
<p>Train a model for matching future typing to profile</p>
</li>
</ol>
<p><strong>Tech Stack:</strong></p>
<ul>
<li>
<p>Python</p>
</li>
<li>
<p><code inline="">pynput</code> or <code inline="">keyboard</code> for keylogging</p>
</li>
<li>
<p><code inline="">scikit-learn</code> or <code inline="">PyTorch</code> for ML</p>
</li>
<li>
<p><code inline="">matplotlib</code> for metrics visualization</p>
</li>
</ul>
<hr>
<p><strong>Full Name:</strong><br>
Shivansh Katiyar</p>
<p><strong>Participant Role:</strong><br>
SSOC</p>
<hr>
</body></html>

### CODE REVIEW

**Comment by SK8-infi** [comment]
The security error is for a password used in tests. And poses no security issues
